src/libraries/System.Security.Cryptography.X509Certificates/tests/MatchesHostnameTests.cs

```diff
@@ -0,0 +1,758 @@
+                    AssertMatch(false, cert, "*");
+                    AssertMatch(false, cert, "*.");
+                    AssertMatch(false, cert, "");
+                }
+            }
+        }
+
+        [Fact]
+        public static void NoMatchOnDnsNameWithLeadingPeriod()
+        {
+            using (ECDsa key = ECDsa.Create())
+            {
+                CertificateRequest req = new CertificateRequest(
+                    "CN=potato.vegetable.example",
+                    key,
+                    HashAlgorithmName.SHA256);
+
+                req.CertificateExtensions.Add(
+                    new X509Extension(
+                        "2.5.29.17",
+                        "301682142E70656163682E66727569742E6578616D706C65".HexToByteArray(),
+                        false));
+
+                DateTimeOffset now = DateTimeOffset.UtcNow;
+                DateTimeOffset notBefore = now.AddMinutes(-1);
+                DateTimeOffset notAfter = now.AddMinutes(1);
+
+                using (X509Certificate2 cert = req.CreateSelfSigned(notBefore, notAfter))
+                {
+                    X509SubjectAlternativeNameExtension sanExt =
+                        (X509SubjectAlternativeNameExtension)cert.Extensions[0];
+
+                    if (sanExt.EnumerateDnsNames().Single() != ".peach.fruit.example")
+                    {
+                        throw new InvalidOperationException("Invalid test data");
+                    }
+
+                    AssertMatch(false, cert, "peach.fruit.example");
+                    AssertMatch(false, cert, "");
+                }
+            }
+        }
+
+        [Fact]
+        public static void WildcardRequiresSuffixToMatch()
+        {
+            using (ECDsa key = ECDsa.Create())
+            {
+                CertificateRequest req = new CertificateRequest(
+                    "CN=potato.vegetable.example",
+                    key,
+                    HashAlgorithmName.SHA256);
+
+                SubjectAlternativeNameBuilder sanBuilder = new SubjectAlternativeNameBuilder();
+                sanBuilder.AddDnsName("*");
+                sanBuilder.AddDnsName("*.");
+                sanBuilder.AddEmailAddress("it@fruit.example");
+
+                req.CertificateExtensions.Add(sanBuilder.Build());
+
+                DateTimeOffset now = DateTimeOffset.UtcNow;
+                DateTimeOffset notBefore = now.AddMinutes(-1);
+                DateTimeOffset notAfter = now.AddMinutes(1);
+
+                using (X509Certificate2 cert = req.CreateSelfSigned(notBefore, notAfter))
+                {
+                    AssertMatch(false, cert, "example");
+                    AssertMatch(false, cert, "example.");
+                    AssertMatch(false, cert, ".");
+                    AssertMatch(true, cert, "*");
+                    AssertMatch(true, cert, "*.");
+                }
+            }
+        }
+
+        [Fact]
+        public static void TooManySANsThrows()
```

Comment:
I didn't quite get what this was testing at first. I thought it was some upper bound on the number of SANs that could be in a single extension.

```suggestion
        public static void TooManySANExtensionsThrows()
```